burstphoto/align.metal

```diff
@@ -94,31 +95,200 @@ kernel void avg_pool(texture2d<float, access::read> in_texture [[texture(0)]],
+    float bw[9] = {1, 0, 0, 0, 0, 0, 0, 0, 0};
+    
+    if(kernel_size_c == 1)      {bw[0]=    2; bw[1]=    1;}
+    else if(kernel_size_c == 2) {bw[0]=    6; bw[1]=    4; bw[2]=   1;}
+    else if(kernel_size_c == 3) {bw[0]=   20; bw[1]=   15; bw[2]=   6; bw[3]=   1;}
+    else if(kernel_size_c == 4) {bw[0]=   70; bw[1]=   56; bw[2]=  28; bw[3]=   8; bw[4]=   1;}
+    else if(kernel_size_c == 5) {bw[0]=  252; bw[1]=  210; bw[2]= 120; bw[3]=  45; bw[4]=  10; bw[5]=  1;}
+    else if(kernel_size_c == 6) {bw[0]=  924; bw[1]=  792; bw[2]= 495; bw[3]= 220; bw[4]=  66; bw[5]= 12; bw[6]=  1;}
+    else if(kernel_size_c == 7) {bw[0]= 3432; bw[1]= 3003; bw[2]=2002; bw[3]=1001; bw[4]= 364; bw[5]= 91; bw[6]= 14; bw[7]= 1;}
+    else if(kernel_size_c == 8) {bw[0]=12870; bw[1]=11440; bw[2]=8008; bw[3]=4368; bw[4]=1820; bw[5]=560; bw[6]=120; bw[7]=16; bw[8]=1;}
+    
+    // compute a sigle output pixel
+    float total_intensity = 0;
+    float total_weight = 0;
+    int x = gid.x;
+    
+    for (int dy = -kernel_size_c; dy <= kernel_size_c; dy++) {
+        int y = gid.y + mosaic_pattern_width*dy;
+        if (0 <= y && y < texture_height) {
+           
+            float const weight = bw[abs(dy)];
+            total_intensity += weight * in_texture.read(uint2(x, y)).r;
+            total_weight += weight;
+        }
+    }
+    
+    // write output pixel
+    float out_intensity = total_intensity / total_weight;
+    out_texture.write(out_intensity, gid);
+}
+
+
+kernel void extend_texture(texture2d<float, access::read> in_texture [[texture(0)]],
+                           texture2d<float, access::write> out_texture [[texture(1)]],
+                           constant int& pad_left [[buffer(0)]],
+                           constant int& pad_top [[buffer(1)]],
+                           uint2 gid [[thread_position_in_grid]]) {
+        
+    int x = gid.x + pad_left;
+    int y = gid.y + pad_top;
+ 
+    float color_value = in_texture.read(gid).r;
+    out_texture.write(color_value, uint2(x, y));
+}
+
+
+kernel void crop_texture(texture2d<float, access::read> in_texture [[texture(0)]],
+                         texture2d<float, access::write> out_texture [[texture(1)]],
+                         constant int& pad_left [[buffer(0)]],
+                         constant int& pad_top [[buffer(1)]],
+                         uint2 gid [[thread_position_in_grid]]) {
+      
+    int x = gid.x + pad_left;
+    int y = gid.y + pad_top;
+  
+    float color_value = in_texture.read(uint2(x, y)).r;
+    out_texture.write(color_value, gid);
+}
+
+
+kernel void add_texture(texture2d<float, access::read> in_texture [[texture(0)]],
+                        texture2d<float, access::read_write> out_texture [[texture(1)]],
+                        constant int& n_textures [[buffer(0)]],
+                        uint2 gid [[thread_position_in_grid]]) {
+    
+    float color_value = out_texture.read(gid).r + in_texture.read(gid).r/float(n_textures);
```

Comment:
This conversion to float can probably be done outside of this function so that we don't pay the penalty of the conversion on every call.

Reply:
I do not think that this change has a measurable impact on performance although it is fully correct. It can be tested in a future version of the app.

Reply:
Resolved in commit [Improved temporal averaging](https://github.com/chris-rank/hdr-plus-swift/commit/4c601bd8bb7af4daf4bdb1a49113a4a7075ef59c).